### DISCUSSION TITLE
Is there support for live resizing?

### DISCUSSION THREAD

**Discussion by sfsam** [rust-windowing/winit#2774] [Q&A] [answered]
On MacOS, there are methods on `NSWindowDelegate` supporting live resizing: 
- `-windowWillStartLiveResize:`
- `-windowDidEndLiveResize:`

Does `winit` have support for these? They could be used, for example, to provide feedback to the user about the window size during resizing or to optimize drawing while live resizing was happening.

**Comment by madsmtm** [comment]
`winit` does not have support for those two, and I'm not sure I understand the benefit of them?

> to provide feedback to the user about the window size during resizing

I'm pretty sure that is done already by `-[NSWindowDelegate windowWillResize:]`/`-[NSView frameDidChange:]`?

> to optimize drawing while live resizing was happening

You mean doing a lower-quality render because the user is less likely to be watching the contents of the window? Or waiting with doing layout until the resize is done?

**Reply by sfsam** [reply]
The ones I am referring to are specifically for when the user does live resizing. There is code in `winit` already for these:
https://github.com/rust-windowing/winit/blob/249609889029af3e5ec19afbd02464e11a265378/src/platform_impl/macos/window_delegate.rs#L121
https://github.com/rust-windowing/winit/blob/249609889029af3e5ec19afbd02464e11a265378/src/platform_impl/macos/window_delegate.rs#L132
But they don't seem to emit events and I was wondering why. Since `winit` is cross-platform, isn't this just a matter of supporting platform functionality? Why wouldn't it?

**Reply by madsmtm** [reply]
Yeah, I know that they're handled by `winit` for internal reasons, I meant what is the use-case for _you_? Or stated alternatively, why would a user need them to be exposed?

**Comment by sfsam** [comment]
For now, my use case is to show the user feedback (dynamically updating dimensions) while they are live resizing a window. But who knows in the future? People use this for a bunch of things beyond what I (or any single person) can think of. See here: https://github.com/search?q=windowwillstartliveresize&type=code.

The reason to expose it is simply because macOS does and that's what cross-platform means. It means a Mac developer can adopt `winit` confident that they can use it to make proper Mac apps without being arbitrarily crippled in some way.

Is there a reason `winit` wouldn't emit these events that are part of the macOS windowing API? The code is practically there already.

**Comment by kchibisov** [comment, marked as answer]
cross platform means providing the same level of experience across multiple platforms. Having backend specific events could be confusing in the first place, especially when it's wrt resizing.

To expose macOS's event in cross platform way, at least all desktop platforms should be investigated on how to do that and a design suggested.

If you develop solely for macOS and don't want to unify stuff like that, use macOS APIs directly (with objc-2 or maybe higher level crates).
